1. ~~To create a Jupyter .ipynb notebook for data analysis~~
2. ~~Import your data or data files and to save as dataframes~~
3. ~~Examine your data, columns and rows and rename and adjust indexing and encoding as appropriate~~
4. ~~Clean null and blank values, and consider to drop rows, as well as to manipulate data and adjust data types as appropriate, including dates and time, or setting appropriate indices. Adjusting specific values and replacing strings and characters for the data wrangling process.~~
5. Explore analysis with graphing and visualizations with matplotlib and seaborn and alternative visualization packages (Plotly, bokeh, altair, vincent)
6. ~~Perform additional analysis by creating new columns for calculations, including aggregator functions, counts and groupbys.~~
7. Encode categorical variables with a variety of techniques through logical conditions, where clauses, or one hot encoding
8. Re-run calculations, including crosstabs or pivots, and new graphs to see results
9. Create correlation matrices, pairplots and heatmaps to determine which attributes should be features for your models and which attributes should not
10. Identify the response variables(s) that you would want to predict/classify/interpret with data science
11. ~~Perform additional feature engineering as necessary, including Min/Max, Normalizaton, Scaling, and additional Pipeline changes that may be beneficial or helpful when you run machine learning~~
12. ~~Merge or concatenate datasets if you have not already, based on common keys or unique items for more in-depth analysis~~
13. ~~Add commenting and markdown throughout the jupyter notebook to explain the interpretation of your results or to comment on code that may not be human readable, and help you recall for you what you are referencing.~~ 
14. To create a markdown .md milestone report that shows and explains the results of what you have accomplished to date in this part of your course project. Consider also creating a .pdf or .pptx to display initial results, aha moments, or findings that would be novel or fascinating for your final presentations.


# Part 3: Exploratory Data Analysis
Project: New Coffee Shop Location
<br/>

## Import data files and save as dataframes.

In [249]:
import pandas as pd
from matplotlib import pyplot as plt

%matplotlib inline

#### First I'll import zip_codes file (which contains all zip codes in dc-metro) into a python list.

In [250]:
import csv
with open('./data/zip_codes.csv', 'r') as f:
    reader = csv.reader(f)
    zip_codes = list(reader)
    
# convert list of lists attained from csv.reader to single flat list
zip_codes_flat = [item for sublist in zip_codes for item in sublist]

#### Import ACS Population Dataset

In [251]:
popul2015_df = pd.read_csv('./data/ACS_15_5YR_B01003_with_ann.csv', names=('zip_code', 'estimate_total'), skiprows=1)
popul2016_df = pd.read_csv('./data/ACS_16_5YR_B01003_with_ann.csv', names=('zip_code', 'estimate_total'), skiprows=1)
populchange_df = popul2015_df.merge(popul2016_df, on="zip_code", suffixes=('_2015', '_2016'))

populchange_df.head()

zip_code  estimate_total_2015  estimate_total_2016
0     20001                43731                44998
1     20002                57412                59659
2     20003                28538                29071
3     20004                 1667                 1708
4     20005                12947                13063

In [216]:
populchange_df['zip_code'] = populchange_df['zip_code'].astype(str)

In [217]:
populchange_df.dtypes

zip_code               object
estimate_total_2015     int64
estimate_total_2016     int64
dtype: object

#### Import Yelp dataset.
    - After importing the dataset I found it to be missing all data from DC metro area, so I chose to use Yelp API.

In [56]:
# businesses = pd.read_json('./data/business.json', lines=True)
# businesses.shape
# businesses.dtypes
# businesses.head()
# businesses[businesses.city.isin(zip_codes_flat)]

In [57]:
# This takes a few minutes to run to run
# Query the yelp api for each zip code in the DC metro area

import requests

url = "https://api.yelp.com/v3/businesses/search"

dict_responses = {}

for i in zip_codes_flat:
    querystring = {'location': i}

    headers = {
        'authorization': "Bearer -Vxv2h6JQ4BvtAFJPnXnFESk2A0ZGIM-Uplb3lm3HbZ0fvyJqyzOaebRuahcJxtKTqdep7oi6ZAyOcHOLU9t0KMp9SK5NZS-TYgmQeC0mjYXRq2XWM0acx31HfMPW3Yx",
        'cache-control': "no-cache",
        'postman-token': "46f50419-2ee9-b8f7-52ec-4f53e3173476"
        }

    response = requests.request("GET", url, headers=headers, params=querystring)
    
    response_todict = response.json()
    response_todict[i] = response_todict.pop('businesses')
    
    dict_responses.update(response_todict)

In [58]:
# Clean up the response from the api

list_of_businesses = []

for key, item in dict_responses.items():
    list_of_businesses.append(item)

del(list_of_businesses[0:2])

# more cleanup; list is currently list of lists of dictionaries due to how api was called
# convert to just list of dictionaries

list_of_businesses_formatting = []

for i in list_of_businesses:
    for j in i:
        list_of_businesses_formatting.append(j)

#### I need to normalize the json data so that each data point is in its own column. This command returns a dataframe.

In [61]:
# Change formatting of categories column to only include "title", not "alias" (makes normalization easier)
for idx, val in enumerate(list_of_businesses_formatting):
    list_categories = []
    for idx2, val2 in enumerate(val['categories']):
        list_categories.append(list_of_businesses_formatting[idx]['categories'][idx2]['title'])
    val['categories'] = list_categories

In [62]:
# This isn't technically normalized all the way since categories still has multiple values.
# Good enough for analysis I need
from pandas.io.json import json_normalize
businesses_df = json_normalize(list_of_businesses_formatting)

## Examine and Clean Data

### I'll start with the Yelp Dataset.

In [63]:
businesses_df.head()

alias  \
0          zaytinya-washington   
1  old-ebbitt-grill-washington   
2   a-baked-joint-washington-9   
3      le-diplomate-washington   
4            rasika-washington   

                                          categories  coordinates.latitude  \
0                         [Greek, Turkish, Lebanese]             38.899040   
1  [Bars, American (Traditional), Breakfast & Bru...             38.898005   
2     [Coffee & Tea, Breakfast & Brunch, Sandwiches]             38.902411   
3                        [Brasseries, French, Cafes]             38.911359   
4                                           [Indian]             38.895008   

   coordinates.longitude   display_phone     distance                      id  \
0             -77.023490  (202) 638-0800   853.458542  GBkFa8TJwkaUJsJXXGkTTg   
1             -77.033362  (202) 347-4800  1428.775801  iyBbcXtQSBfiwFQZwVBNaQ   
2             -77.017139  (202) 408-6985   547.273915  SpCeYPhky4gsWa9-IBtw2A   
3             -77.031575  (202) 332-3333  1086.375403  j9qYRR8HCXm_GEnetijOGA   
4             -77.021286  (202) 637-1222  1267.482230  CwdlygqT4cWwOtQGsYdoBw   

                                           image_url  is_closed  \
0  https://s3-media2.fl.yelpcdn.com/bphoto/qf8Mc1...      False   
1  https://s3-media2.fl.yelpcdn.com/bphoto/KBCezp...      False   
2  https://s3-media1.fl.yelpcdn.com/bphoto/iTBw1K...      False   
3  https://s3-media2.fl.yelpcdn.com/bphoto/2EljPz...      False   
4  https://s3-media4.fl.yelpcdn.com/bphoto/rkzs8J...      False   

  location.address1                        ...                          \
0     701 9th St NW                        ...                           
1    675 15th St NW                        ...                           
2       440 K St NW                        ...                           
3   1601 14th St NW                        ...                           
4       633 D St NW                        ...                           

                  location.display_address location.state location.zip_code  \
0    [701 9th St NW, Washington, DC 20001]             DC             20001   
1   [675 15th St NW, Washington, DC 20005]             DC             20005   
2      [440 K St NW, Washington, DC 20001]             DC             20001   
3  [1601 14th St NW, Washington, DC 20009]             DC             20009   
4      [633 D St NW, Washington, DC 20004]             DC             20004   

               name         phone price rating review_count transactions  \
0          Zaytinya  +12026380800   $$$    4.0         3944           []   
1  Old Ebbitt Grill  +12023474800    $$    4.0         6536           []   
2     A Baked Joint  +12024086985     $    4.5         1215           []   
3      Le Diplomate  +12023323333   $$$    4.0         2448           []   
4            Rasika  +12026371222   $$$    4.5         2631           []   

                                                 url  
0  https://www.yelp.com/biz/zaytinya-washington?a...  
1  https://www.yelp.com/biz/old-ebbitt-grill-wash...  
2  https://www.yelp.com/biz/a-baked-joint-washing...  
3  https://www.yelp.com/biz/le-diplomate-washingt...  
4  https://www.yelp.com/biz/rasika-washington?adj...  

[5 rows x 24 columns]

In [69]:
businesses_df.shape

(6946, 24)

#### My looping api call technique resulted in many duplicates, as we examine and fix below (id is unique).

In [70]:
businesses_df.id.value_counts().head()

VUfflugAZa3MxbtzDTGnEA    20
iyBbcXtQSBfiwFQZwVBNaQ    19
VA8aPObRynlwR1TGzbzraQ    19
GBkFa8TJwkaUJsJXXGkTTg    15
CwdlygqT4cWwOtQGsYdoBw    14
Name: id, dtype: int64

In [71]:
businesses_df.drop_duplicates(subset='id', inplace=True)

#### Now I will drop columns that I won't need during this project.

In [73]:
businesses_df.drop(['alias',
         'id',
         'display_phone',
         'distance', 
         'image_url', 
         'is_closed', 
         'location.address2', 
         'location.address3', 
         'location.display_address',
         'phone',
         'transactions',
         'url'], axis=1, inplace=True)

#### Next I'll clean up the column names.

In [74]:
businesses_df.columns = ['categories', 'latitude', 
                         'longitude', 'address', 
                         'city', 'country', 'state', 
                         'zip_code', 'name', 'price', 
                         'rating', 'review_count']

In [75]:
# move name column to first position
cols = businesses_df.columns.tolist()
cols[0] = 'name'
cols[8] = 'categories'
businesses_df = businesses_df[cols]

In [76]:
businesses_df.head()

name   latitude  longitude          address            city  \
0          Zaytinya  38.899040 -77.023490    701 9th St NW  Washington, DC   
1  Old Ebbitt Grill  38.898005 -77.033362   675 15th St NW  Washington, DC   
2     A Baked Joint  38.902411 -77.017139      440 K St NW  Washington, DC   
3      Le Diplomate  38.911359 -77.031575  1601 14th St NW  Washington, DC   
4            Rasika  38.895008 -77.021286      633 D St NW  Washington, DC   

  country state zip_code                                         categories  \
0      US    DC    20001                         [Greek, Turkish, Lebanese]   
1      US    DC    20005  [Bars, American (Traditional), Breakfast & Bru...   
2      US    DC    20001     [Coffee & Tea, Breakfast & Brunch, Sandwiches]   
3      US    DC    20009                        [Brasseries, French, Cafes]   
4      US    DC    20004                                           [Indian]   

  price  rating  review_count  
0   $$$     4.0          3944  
1    $$     4.0          6536  
2     $     4.5          1215  
3   $$$     4.0          2448  
4   $$$     4.5          2631

#### Check for relevent null values

In [77]:
businesses_df.isnull().sum()

name              0
latitude          0
longitude         0
address           5
city              0
country           0
state             0
zip_code          0
categories        0
price           228
rating            0
review_count      0
dtype: int64

In [79]:
businesses_df[businesses_df.address.isnull()]

name   latitude  longitude address            city  \
1656           Rolling Cow  38.898260 -77.042200    None  Washington, DC   
3370           Pizza Brama  39.056313 -77.161659    None         Potomac   
3375           NeatMeat DC  38.950391 -77.015989    None  Washington, DC   
3438           La Pote-Ria  39.238080 -77.278570    None      Clarksburg   
3450  Dyvine BBQ In Motion  38.565160 -77.325150    None        Dumfries   

     country state zip_code                                     categories  \
1656      US    DC    20006  [Food Trucks, Korean, American (Traditional)]   
3370      US    MD    20854                           [Pizza, Food Trucks]   
3375      US    DC                           [Food Trucks, American (New)]   
3438      US    MD                                [Food Trucks, Colombian]   
3450      US    VA    22026              [Food Trucks, Caterers, Barbeque]   

     price  rating  review_count  
1656    $$     4.5            51  
3370    $$     5.0            16  
3375    $$     4.0            38  
3438     $     4.5            27  
3450     $     4.5            17

In [83]:
businesses_df[businesses_df.zip_code==''].zip_code.value_counts()

    8
Name: zip_code, dtype: int64

In [96]:
# There are rows missing zip codes. 
# Seeing as zip code is an important factor in my hypotheses, I'll drop these rows.
businesses_df = businesses_df[businesses_df.zip_code != '']

### Next, ACS dataset.
Pretty straightforward and clean.

In [254]:
populchange_df.sort_values('zip_code', ascending=False).head()

zip_code  estimate_total_2015  estimate_total_2016
356     25446                 1638                 1840
355     25443                 7442                 7439
354     25442                 1519                 1748
353     25438                 6932                 6741
352     25432                   71                  112

In [252]:
populchange_df.isnull().sum()

zip_code               0
estimate_total_2015    0
estimate_total_2016    0
dtype: int64

## Explore Data

#### First I want to find a count of coffee shops by zip code.

In [248]:
businesses_df[businesses_df.categories.astype(str).str.contains('Coffee')].shape

(327, 12)

In [178]:
businesses_df[businesses_df.name.astype(str).str.contains('Starbucks')].name.value_counts()

Starbucks    20
Name: name, dtype: int64

#### After starting to explore the data I realized the Yelp dataset was for some reason missing several starbucks locations. I went to an external dataset to bring these in, and then append them to the original businesses dataset.

In [241]:
starbucks_df = pd.read_csv('./data/starbucks_locations.csv')

# filter out starbucks locations not in one of the relevent zip codes
starbucks_df = starbucks_df[starbucks_df['Country'].isin(['US']) & starbucks_df['Postcode'].isin(zip_codes_flat)]
starbucks_df.head(2)

Brand  Store Number                   Store Name Ownership Type  \
16073  Starbucks  23852-232305  Giant - Washington DC #2379       Licensed   
16079  Starbucks  10439-100821          500 New Jersey Ave.  Company Owned   

                 Street Address        City State/Province Country Postcode  \
16073     3336 Wisconsin Ave NW  Washington             DC      US    20016   
16079  500 New Jersey Avenue NW  Washington             DC      US    20001   

         Phone Number                    Timezone  Longitude  Latitude  
16073    202-237-5820  GMT-05:00 America/New_York     -77.07     38.93  
16079  (202) 347-0364  GMT-05:00 America/New_York     -77.01     38.90

In [242]:
# drop irrelevent columns, rename remaining columns, add categories column to new Starbucks dataset
starbucks_df.drop(['Store Number', 'Store Name', 'Ownership Type', 'Timezone', 'Phone Number'], axis=1, inplace=True)
starbucks_df.columns=['name', 'address', 'city', 'state', 'country', 'zip_code', 'longitude', 'latitude']
starbucks_df['categories'] = '[Coffee]'

In [244]:
# drop starbucks locations already in businesses_df to avoid duplicates
starbucks_drop_list = businesses_df[businesses_df.name.astype(str).str.contains('Starbucks')].index.values
businesses_df.drop(starbucks_drop_list, inplace=True)

# add starbucks locations to businesses
businesses_df = businesses_df.append(starbucks_df)

#### Explore population data viability on determining coffee shop location

In [245]:
# get all businesses w/ category containing "Coffee"
coffeeshops_byzip_df = businesses_df[businesses_df.categories.astype(str).str.contains('Coffee')]
coffeeshops_byzip_df.shape

(327, 12)

In [246]:
# new df containing each zip code and the count of coffee shops
coffeeshop_percap_df = coffeeshops_byzip_df.groupby('zip_code').zip_code.agg(['count']).sort_values('count', ascending=False)

# merge previous dataframe with population growth dataframe
coffeeshop_percap_df = coffeeshop_percap_df.merge(populchange_df, left_index=True, right_on="zip_code")
coffeeshop_percap_df = coffeeshop_percap_df.set_index('zip_code')

# add calculated field for rate of population change
coffeeshop_percap_df['pop_change_rate'] = (coffeeshop_percap_df['estimate_total_2016'] - \
                                            coffeeshop_percap_df['estimate_total_2015']) / \
                                            coffeeshop_percap_df['estimate_total_2015']

# display some rows from newly constructed df, sorted by population rate of change
coffeeshop_percap_df[coffeeshop_percap_df['count'] > 5].sort_values('pop_change_rate')

count  estimate_total_2015  estimate_total_2016  pop_change_rate
zip_code                                                                  
20166         8                10150                 9961        -0.018621
22003         7                60058                60039        -0.000316
20006        10                 2905                 2920         0.005164
20003         8                28538                29071         0.018677
20001        11                43731                44998         0.028973
20002         6                57412                59659         0.039138
22202         7                22337                23379         0.046649

#### Verdict on population...
Looking at area code 20006, I'm starting to think population is not a good way to predict a good coffee shop location, especially within DC (20006 (including white house) contains tourist attractions, shopping, places of employment, etc; not much housing). What I am missing is a dataset that tells me what people are doing/where they're coming from when they go for coffee, but I'd predict they're not usually coming from home. More likely people would be coming from work, shopping, etc.